nimare/annotate/gclda.py

```diff
@@ -14,6 +15,247 @@
+        ny_region_topic[region, topic] -= 1  # Decrement count in Subregion x Topic count matx
+        ny_doc_topic[doc, topic] -= 1  # Decrement count in Document x Topic count matrix
+
+        # Retrieve the probability of generating current x from all
+        # subregions: [R x T] array of probs
+        p_x_subregions = (peak_probs[i_ptoken, :, :]).transpose()
+
+        # Compute the probabilities of all subregions given doc
+        #     p(r|d) ~ p(r|t) * p(t|d)
+        # Counts of subregions per topic + prior: p(r|t)
+        p_region_g_topic = ny_region_topic + delta
+
+        # Normalize the columns such that each topic's distribution over subregions sums to 1
+        p_region_g_topic_sum = np.sum(p_region_g_topic, axis=0)
+        p_region_g_topic = np.divide(p_region_g_topic, p_region_g_topic_sum)
+
+        # Counts of topics per document + prior: p(t|d)
+        p_topic_g_doc = ny_doc_topic[doc, :] + alpha
+
+        # Reshape from (ntopics,) to (nregions, ntopics) with duplicated rows
+        p_topic_g_doc = p_topic_g_doc.repeat(n_regions).reshape((-1, n_regions)).T
+
+        # Compute p(subregion | document): p(r|d) ~ p(r|t) * p(t|d)
+        # [R x T] array of probs
+        p_region_g_doc = p_topic_g_doc * p_region_g_topic
+
+        # Compute the multinomial probability: p(z|y)
+        # Need the current vector of all z and y assignments for current doc
+        # The multinomial from which z is sampled is proportional to number
+        # of y assigned to each topic, plus constant gamma
+        # Compute the proportional probabilities in log-space
+        logp = nz_doc_topic[doc, :] * np.log(
+            (ny_doc_topic[doc, :] + gamma + 1) / (ny_doc_topic[doc, :] + gamma)
+        )
+        # Add a constant before exponentiating to avoid any underflow issues
+        p_peak_g_topic = np.exp(logp - np.max(logp))
+
+        # Reshape from (ntopics,) to (nregions, ntopics) with duplicated rows
+        p_peak_g_topic = p_peak_g_topic.repeat(n_regions).reshape((-1, n_regions)).T
+
+        # Get the full sampling distribution:
+        # [R x T] array containing the proportional probability of all y/r combinations
+        probs_pdf = p_x_subregions * p_region_g_doc * p_peak_g_topic
+
+        # Convert from a [R x T] matrix into a [R*T x 1] array we can sample from
+        probs_pdf = np.reshape(probs_pdf, (n_regions * n_topics))
+
+        # Normalize the sampling distribution
+        probs_pdf = probs_pdf / np.sum(probs_pdf)
+        # Float precision issue in Numba: https://github.com/numba/numba/issues/3426
+        probs_pdf = np.trunc(probs_pdf * (10**12)) / (10**12)
+
+        # Sample a single element (corresponding to a y_i and c_i assignment for the ptoken)
+        # from the sampling distribution
+        # Returns a binary [1 x R*T] vector with a '1' in location that was sampled
+        # and zeros everywhere else
+        assignment_vec = np.random.multinomial(1, probs_pdf)
```

Comment:
The documentation says to "use the `multinomial` method of a `default_rng()` instance instead". If we're lucky, that might also be faster?

Reply:
I have tried different combinations with:
```
rng = np.random.default_rng()
assignment_vec = rng.multinomial(1, probs_pdf)
```
and everything has failed in numba.

Using the correction proposed in this issue `https://github.com/numba/numba/issues/3426`, the current `assignment_vec = np.random.multinomial(1, probs_pdf)` is working fine without truncating `probs_pdf`.

Reply:
Ah that's good to hear. I'd love to hear what the speed up is using numba on a large-ish dataset.